deterministic_kernel.hpp

```diff
@@ -0,0 +1,246 @@
+ * cdf returns the upper range that encompasses x percent of the distribution (e.g for 99% input .99)
+ */
+class CauchyDistribution
+{
+public:
+	CauchyDistribution(double scale, double locator)
+        : s(scale), t(locator)
+    {}
+
+    double pdf(double x)
+    {
+    	return 1 / ((s*PI)*(1 + (pow((x - t)/s, 2))));
+    
+    }
+    
+    double cdf(double x)
+    {
+    	if ( s == 1 && t == 0) {
+    		// checking upper half so multiply result by 2
+    		return tan((x / 2) * M_PI) * 2;
+    	} else {
+    		return (tan((x/2)*M_PI + (atan(-t/s))) * s + t) * 2;
+    	}
+    }
+private:
+	// scale parameter - 1 for standard
+    double s;
+    // location parameter - location of peak - 0 for standard
+    double t;
+};
+
+/*! 
+ * Exponential distribution
+ * Includes probability density function and cumulative distribution function
+ * pdf returns the probability that the variate has the value x
+ * cdf returns the upper range that encompasses x percent of the distribution (e.g for 99% input 0.99)
+ */
+class ExponentialDistribution
+{
+public:
+ExponentialDistribution(double scale)
+        : beta(scale)
+    {}
+    // assumes mu is 0 which is traditionally accepted
+    double pdf(double x)
+    {
+    	return (1/beta)*(exp(-x/beta));
+    }
+    
+    double cdf(double x)
+    {
+    	if ( beta == 1) {
+    		return -log(1-x);
+    	} else {
+    		return -beta * log(1-x);
+    	}
+    }
+private:
+	// scale parameter - 1 for standard 
+	// equal to 1/lambda
+	double beta;
+};
+
+/*! 
+ * Dispersal kernel for deterministic spread to cell with highest probability of spread
+ *
+ * Dispersal Kernel type determines use of Exponential or Cauchy distribution
+ * to find probability.
+ *
+ * dispersal_percentage is the percent of all possible dispersal to be included 
+ * in the moving window size (e.g for 99% input 0.99).
+ *
+ * Useful for testing as it is deterministic and provides fully replicable results 
+ */
+class DeterministicDispersalKernel
+{
+protected:
+    Raster<int> dispersers_;
+    DispersalKernelType kernel_type_;
+    // the west-east resolution of the pixel
+    double east_west_resolution;
+    // the north-south resolution of the pixel
+    double north_south_resolution;
+    // row/col position of middle cell
+    int mid_row = 0;
+    int mid_col = 0;
+    // position of cell from previous call
+    int prev_row = -1;
+    int prev_col = -1;
+    // number of rows/cols in the probability window
+    int number_of_rows = 0;
+    int number_of_columns = 0;
+    // maximum distance from center cell to outer cells
+    double max_distance;
+    std::vector<std::vector<double>> probability;
+    std::vector<std::vector<double>> probability_copy;
+    CauchyDistribution cauchy;
+    ExponentialDistribution exponential;
+    double number_of_dispersers;
+public:
+    DeterministicDispersalKernel(DispersalKernelType dispersal_kernel, 
+    							Raster<int> dispersers, double dispersal_percentage,
```

Comment:
The type of dispersers needs to be little more complicated:

1. It should be a const reference, i.e., `const ... &` so that the data does not have to be copied, but it is guaranteed in the API that we don't modify the data.
2. The kernel class actually needs to be a class template with one template parameter IntegerRaster, see, e.g., the Simulation class. This allows us to use different raster/matrices/grids in the different APIs. If this turns out to be too much hassle, I can add this generalization later.